Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Kms/KmsXmlEncryptor.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright 2019 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.Kms.V1;
+using Google.Protobuf;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.DataProtection.XmlEncryption;
+using System.IO;
+using System.Xml.Linq;
+
+namespace Google.Cloud.AspNetCore.DataProtection.Kms
+{
+    /// <summary>
+    /// Implementation for encryption, configured by <see cref="GoogleCloudDataProtectionBuilderExtensions"/>
+    /// </summary>
+    internal sealed class KmsXmlEncryptor : IXmlEncryptor
+    {
+        private readonly KeyManagementServiceClient _kmsClient;
+        private readonly CryptoKeyName _keyName;
+        private readonly CryptoKeyPathName _keyPathName;
+
+        internal KmsXmlEncryptor(KeyManagementServiceClient kmsClient, CryptoKeyName keyName)
+        {
+            _kmsClient = GaxPreconditions.CheckNotNull(kmsClient, nameof(kmsClient));
+            _keyName = GaxPreconditions.CheckNotNull(keyName, nameof(keyName));
+            _keyPathName = new CryptoKeyPathName(keyName.ProjectId, keyName.LocationId, keyName.KeyRingId, keyName.CryptoKeyId);
+        }
+   
+        /// <inheritdoc />
+        public EncryptedXmlInfo Encrypt(XElement plaintextElement)
+        {
+            ByteString plainTextData;
+
+            using (var stream = new MemoryStream())
+            {
+                plaintextElement.Save(stream, SaveOptions.DisableFormatting);
+                stream.Position = 0;
+                plainTextData = ByteString.FromStream(stream);
+            }
+
+            ByteString encryptedData = _kmsClient.Encrypt(_keyPathName, plainTextData).Ciphertext;
```

Comment:
When plaintext data > 64k, the call to KMS will fail.  In that case, we need to use envelope encryption:
https://cloud.google.com/kms/docs/envelope-encryption

Reply:
I believe ASP.NET Core is basically using envelope encryption anyway:

- It generates a key
- It encrypts the key using the IXmlEncryptor
- It stores the key using the IXmlRepository
- All servers can now fetch that encrypted key, and decrypt it to get the *original* key which is used to sign requests

As such, I think we can probably just validate at this point that `plainTextData` is no more than 64K long, and throw an exception (with a really clear message) if it is. While we *could* encrypt it one chunk at a time, I strongly suspect that's a bad idea from a security standpoint.

Good catch.